### PR TITLE
Adding back endpoints tests.

### PR DESCRIPTION
**Pull request by macieksmuga**

### CODE REVIEW

**Comment by macieksmuga** [comment]
I added these back as a separate PR to be consistent with the other test-adding PR's, but note that issue #88 questions the need for these tests as part of the compliance suite.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    private static String makeFullUrl(String partialUrl) {
+        return TransportUtils.makeUrl(urls.getUrlRoot(), partialUrl);
+    }
+
+    /**
+     * Test that the basic server verbs/methods work as expected.
+     *
+     * @param fullUrl the URL to test
+     */
+    private void testHttpMethods(String fullUrl) throws UnirestException {
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: This test is quite opaque to me --- what are we trying to test here?

This might be Java heresy, but it is helpful to break things up into multiple lines sometimes...

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        assertThat(Unirest.get(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_BAD_METHOD);
+        assertThat(Unirest.options(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
+        assertThat(Unirest.post(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_UNSUPPORTED_TYPE);
+    }
+
+    /**
+     * Test the status codes we're supposed to receive from the GET, POST, and OPTIONS methods on
+     * <tt>/callsets/search</tt>.
+     *
+     * <p>See server/tests/unit/test_views.py: testRouteCallsets</p>
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: I don't think we should be citing the server tests, if for no other reason that we refactor pretty aggressively and this file might not exist at all in a few months time.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * Test that the basic server verbs/methods work as expected.
+     *
+     * @param fullUrl the URL to test
+     */
+    private void testHttpMethods(String fullUrl) throws UnirestException {
+        assertThat(Unirest.get(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_BAD_METHOD);
+        assertThat(Unirest.options(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
+        assertThat(Unirest.post(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_UNSUPPORTED_TYPE);
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Why is POST returning HTTP_UNSUPPORTED_TYPE here?

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * Test the status codes we're supposed to receive from the GET, POST, and OPTIONS methods on
+     * <tt>/callsets/search</tt>.
+     *
+     * <p>See server/tests/unit/test_views.py: testRouteCallsets</p>
+     *
+     * @throws Exception if there's a connection problem
+     */
+    @Test
+    public void checkCallSetsRouting() throws Exception {
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: This seems something of an outlier. I'm expecting to see all the endpoints if CallSets are included here.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        testHttpMethods(makeFullUrl(partialUrl));
+    }
+
+    /**
+     * Test that searches' verbs/methods work as expected.
+     *
+     * @param fullUrl the URL to test
+     */
+    private void testSearchRouting(final String fullUrl) throws UnirestException {
+        // send some malformed requests and expect status == HTTP_BAD_REQUEST
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: This is a good test, but should be names something more specific. TestBadJson?

**Comment by jeromekelleher** [comment]
I think the framework is good here and we should definitely have tests like this (we do need to check some top-level error behaviour and so on). However, a lot of the tests here don't make much sense to me. I would pare this down to just testing the BadJson on the search endpoints (for which the expected behaviour is pretty clear), and look to extend these tests over time in conjunction with updating the high-level documentation.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 1 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Neither of those methods is a (JUnit) test: they're both utility methods called by test methods defined elsewhere in the file.

How to tell what's a test method from what's not?  Tests must be:
- annotated with the `@Test` tag.
- declared as `public void`.
- declared to throw some kind of exception, or subclass of exception.

Sorry if the name `testHttpMethods` led you to think otherwise.  If I recall correctly, older versions of JUnit looked for `public void` methods with names beginning with `test`, which probably contributed to the confusion.

The only true JUnit tests in this file are `checkCallSetsRouting`, `checkVariantSearchMethods`, and `checkSearchRouting`.

I'm not sure I understand your comment about multiple lines - can you explain?

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 2 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Your point is well taken.  However, I wanted to provide some provenance for this test, since I couldn't find anything describing the behavior in ga4gh/schemas.

Perhaps it's not standard cross-server behavior at all, and this test should go away?

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 3 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: It's a POST sans payload, and the ref server returns 415 for that.  As above, maybe this is not something we want to require.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 4 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Good point.  It's kind of vestigial.

It's easy enough to generalize this to run across all endpoints.  I will see if there's a reason not to do that.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 5 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Yes, the present name is bad.  I'll come up with something more appropriate.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 1 above)
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Well I guess I was mildly complaining that I find it hard to follow stuff like
```java
 assertThat(Unirest.get(fullUrl).asBinary().getStatus()).isEqualTo(HttpURLConnection.HTTP_BAD_METHOD);
```
What's Unirest? Why do we call asBinary()? To me these are details distracting me from the point of the test, which is to check the http status. It would be easier to read if there was something like
```java
assertEqual(getHttpStatus(url), HttpURLConnection.HTTP_BAD_METHOD);
```
It seems to be the common style here, but I find it slow to parse these long chains of method calls, and generally resent having to put my finger on the screen to trace where brackets begin and end.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 2 above)
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: I think if it's not currently documented (and not obviously the right thing to do) we should remove the test. Opening an issue on it for the schemas so that it is documented, and incorporating the test afterwards is the way to go I think. The baseline should be uncontroversial stuff, and we don't want undocumented assumptions made about the HTTP behaviour becoming requirements.

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 3 above)
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: Again, if it's not documented we shouldn't require the behaviour. (I'm not saying it's not the right behaviour, just that it's not reasonable to make undocumented compliance requirements.)

File: cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java (hunk 1 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/endpoints/RawEndpointsIT.java]: I hope the code I write is generally less dense than that, but I take your point.

**Comment by jeromekelleher** [comment]
Looks good @hjellinek, thanks for the updates.

+1.

**Comment by calbach** [comment]
+1

**Comment by jeromekelleher** [comment]
Looks like everyone is happy with this; merging. Thanks @hjellinek and @macieksmuga!
